perllib/FYR/Queue.pm

```diff
@@ -1135,6 +1144,116 @@ sub email_template_params ($%) {
+    }
+
+    # Convert the plain text message into html so it displays
+    # more or less correctly in HTML emails. This splits the
+    # address portion and the message portion as they have
+    # different formatting requirements
+    if ($html_settings->{email_text}) {
+        my $msg = $html_settings->{email_text};
+        my ($address, $body) = split('Email:', $msg, 2);
+        $address =~ s%\n%<br/>%gs;
+        if ($body =~ /\d\s\w+\s\d{4}\n\s*Dear/s) {
+            my ($email_and_date, $message) = split(/^Dear /m, $body, 2);
+            $email_and_date =~ s%\n\n%</p>\n<p align="right">%s;
+            $message =~ s%\n\n%</p>\n<p>%gs;
+            $html_settings->{email_text} = $address .
+                '</p><p align="right">Email:' .
+                $email_and_date .
+                '</p><p>Dear ' .
+                $message;
+        } else {
+            $body =~ s%\n\n%</p>\n<p>%gs;
+            $html_settings->{email_text} = $address .
+                '</p><p>Email:' .
+                $body;
+        }
+    }
+
+    my $logo = Email::MIME->create(
+       attributes => {
+            filename     => "logo.gif",
+            content_type => "image/gif",
+            encoding     => "base64",
+            name         => "logo.gif",
+       },
+       body => io( email_template('logo.gif', $msg->{cobrand}) )->binary->all
+    );
+    $logo->header_set('Content-ID', '<logo.gif>');
+
+    # bounce emails don't have a message so we don't pass them to
+    # email_template_params as it needs one.
+    my $params = $html_settings;
+    if ( keys %$msg ) {
+        $params = email_template_params($msg, %$html_settings);
+    }
+
+    my $bodyhtml = FYR::EmailTemplate::format(
+                email_template('_top.html', $msg->{cobrand}),
+                $params
+            );
+    $bodyhtml .= FYR::EmailTemplate::format(
+                email_template($template . '.html', $msg->{cobrand}),
+                $params
+            );
+    $bodyhtml .= FYR::EmailTemplate::format(
+                email_template('_bottom.html', $msg->{cobrand}),
+                $params
+            );
+
+    my $html = Email::MIME->create(
+        body_str => $bodyhtml,
+        attributes => {
+            charset => 'utf-8',
+            encoding => 'quoted-printable',
+            content_type => 'text/html'
+        }
+    );
+
+    foreach ($logo, $html) {
+        $_->header_set('Date');
+        $_->header_set('MIME-Version');
+    }
+
+    my $mail = Email::MIME->create(
+        attributes => {
+            charset => 'utf-8',
+            content_type => 'multipart/related'
+        },
+        parts => [ $html, $logo ]
+    );
+
+    $mail->header_set('Date');
+    $mail->header_set('MIME-Version');
+
+    return $mail;
+}
+
+sub build_text_email($) {
+    my $text = shift;
+
+    my $email = Email::MIME->create(
+        body_str => $text,
+        attributes => {
+            charset => 'utf-8',
+            encoding => '7bit',
```

Comment:
This won't work if there any top-bit set characters, I'm afraid, it passes them through without alteration, it needs to be quoted-printable.